File: manual/data/cpp/tests/test_pycrypto_aes.cpp
#include <gtest/gtest.h>
#include "../src/function_pycrypto_aes.cpp"


class TestAESCipher : public ::testing::Test {
protected:
    void SetUp() override {
        key = "mysecretpassword";
        plaintext = "This is a test message for AESCipher.";
        cipher = new AESCipher();
    }

    void TearDown() override {
        delete cipher;
    }

    std::string key;
    std::string plaintext;
    AESCipher* cipher;
};

TEST_F(TestAESCipher, EncryptDecryptBasicText) {
    std::string plaintext = "Hello, World!@#$%^&*()_+-=~1234567890";
    std::string encrypted = cipher->process(plaintext, key, "encrypt");
    std::string decrypted = cipher->process(encrypted, key, "decrypt");
    EXPECT_EQ(decrypted, plaintext);
}

TEST_F(TestAESCipher, DecryptWithWrongKey) {
    std::string plaintext = "This is a test message.";
    std::string encrypted = cipher->process(plaintext, key, "encrypt");
    std::string wrong_key = "wrongpassword";
    EXPECT_THROW(cipher->process(encrypted, wrong_key, "decrypt"), std::runtime_error);
}

TEST_F(TestAESCipher, DecryptPartialCiphertext) {
    std::string plaintext = "Partial decryption test";
    std::string encrypted = cipher->process(plaintext, key, "encrypt");
    std::string partial_encrypted = encrypted.substr(0, encrypted.length() / 2);
    EXPECT_THROW(cipher->process(partial_encrypted, key, "decrypt"), std::runtime_error);
}

TEST_F(TestAESCipher, DecryptTamperedCiphertext) {
    std::string plaintext = "Tamper test";
    std::string encrypted = cipher->process(plaintext, key, "encrypt");
    std::string tampered_encrypted = encrypted.substr(0, encrypted.length()-1) + "X";
    EXPECT_THROW(cipher->process(tampered_encrypted, key, "decrypt"), std::runtime_error);
}

TEST_F(TestAESCipher, AES128_192_256BitKey) {
    std::string plaintext = this->plaintext;

    std::string key_128 = "thisis128bitkey";
    std::string encrypted = cipher->process(plaintext, key_128, "encrypt");
    std::string decrypted = cipher->process(encrypted, key_128, "decrypt");
    EXPECT_EQ(decrypted, plaintext);

    std::string key_192 = "thisis192bitkey______";
    encrypted = cipher->process(plaintext, key_192, "encrypt");
    decrypted = cipher->process(encrypted, key_192, "decrypt");
    EXPECT_EQ(decrypted, plaintext);

    std::string key_256 = "thisis256bitkey______________";
    encrypted = cipher->process(plaintext, key_256, "encrypt");
    decrypted = cipher->process(encrypted, key_256, "decrypt");
    EXPECT_EQ(decrypted, plaintext);
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: manual/data/java/tests/test_pycrypto_aes.java
package com.example;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

public class test_pycrypto_aes {

    private String key = "mysecretpassword";
    private String plaintext = "This is a test message for AESCipher.";
    private function_pycrypto_aes cipher;

    private String encrypted;
    private String decrypted;

    @BeforeEach
    void setUp() {
        cipher = new function_pycrypto_aes();
    }

    @Test
    void test_encrypt_decrypt_basic_text() {
        String plaintext = "Hello, World!@#$%^&*()_+-=~1234567890";
        encrypted = cipher.process(plaintext, key, "encrypt");
        decrypted = cipher.process(encrypted, key, "decrypt");
        assertEquals(plaintext, decrypted);
    }

    @Test
    void test_decrypt_with_wrong_key() {
        String plaintext = "This is a test message.";
        encrypted = cipher.process(plaintext, key, "encrypt");
        String wrongKey = "wrongpassword";
        assertThrows(IllegalArgumentException.class, () -> cipher.process(encrypted, wrongKey, "decrypt"));
    }

    @Test
    void test_decrypt_partial_ciphertext() {
        String plaintext = "Partial decryption test";
        encrypted = cipher.process(plaintext, key, "encrypt");
        String partialEncrypted = encrypted.substring(0, encrypted.length() / 2);
        assertThrows(IllegalArgumentException.class, () -> cipher.process(partialEncrypted, key, "decrypt"));
    }

    @Test
    void test_decrypt_tampered_ciphertext() {
        String plaintext = "Tamper test";
        encrypted = cipher.process(plaintext, key, "encrypt");
        String tamperedEncrypted = encrypted.substring(0, encrypted.length() - 1) + "X";
        assertThrows(IllegalArgumentException.class, () -> cipher.process(tamperedEncrypted, key, "decrypt"));
    }

    @Test
    void test_aes_128_192_256_bit_key() {
        String plaintext = this.plaintext;

        String key128 = "thisis128bitkey";
        encrypted = cipher.process(plaintext, key128, "encrypt");
        decrypted = cipher.process(encrypted, key128, "decrypt");
        assertEquals(plaintext, decrypted);

        String key192 = "thisis192bitkey______";
        encrypted = cipher.process(plaintext, key192, "encrypt");
        decrypted = cipher.process(encrypted, key192, "decrypt");
        assertEquals(plaintext, decrypted);

        String key256 = "thisis256bitkey______________";
        encrypted = cipher.process(plaintext, key256, "encrypt");
        decrypted = cipher.process(encrypted, key256, "decrypt");
        assertEquals(plaintext, decrypted);
    }
}
